net/http2_conn_pool.go

```diff
@@ -0,0 +1,293 @@
+	if atomic.LoadInt64(&pool.numConns) >= int64(pool.maxConns) && pool.maxConns > 0 {
+		pool.Unlock()
+		return false, nil
+	}
+
+	newConn := &connNode{
+		obj: obj,
+	}
+
+	if atomic.LoadInt64(&pool.numConns) == 0 {
+		newConn.next = newConn
+		newConn.prev = newConn
+		pool.currentConn = newConn
+		atomic.AddInt64(&pool.numConns, 1)
+		select {
+		case pool.insertedWhileNoneCh <- struct{}{}:
+		default:
+		}
+
+		pool.Unlock()
+		return true, nil
+	}
+
+	newConn.next = pool.currentConn
+	newConn.prev = pool.currentConn.prev
+	pool.currentConn.prev.next = newConn
+	pool.currentConn.prev = newConn
+	atomic.AddInt64(&pool.numConns, 1)
+
+	pool.Unlock()
+	return true, nil
+}
+
+// populateAvailable is a run loop which constantly updates the channel of
+// connections to be used
+func (pool *connPool) populateAvailable() {
+	for {
+		pool.Lock()
+
+		if pool.currentConn == nil {
+			// create notify chan before unlock to prevent deadlock
+			notify := make(chan interface{})
+			go func() {
+				<-pool.insertedWhileNoneCh
+				close(notify)
+			}()
+
+			// unlock to allow insertion
+			pool.Unlock()
+
+			pool.logger.Warn("No connections in connection pool currently: waiting")
+
+			// wait for insertion
+			<-notify
+
+			pool.logger.Warn("No connections in connection pool currently: got new conn notification")
+			// TODO: add more connections in this case. Some sort of queue or signal system
+			continue
+		}
+
+		// wait until this connection has gone through the round tripper if it's been queued
+		// this is how we make sure we don't send repeat conns down the conn chan without
+		// checking if they can actually serve a request
+		//
+		// TODO: Skip a conn which has been Acquired. A long lasting request could freeze the whole conn pool
+		pool.currentConn.obj.AcquireUsage()
+		if !pool.currentConn.obj.Expended() {
+			retConn := pool.currentConn
+
+			// iterate to next conn for next request
+			pool.currentConn = pool.currentConn.next
+
+			pool.goodConnCh <- retConn
+			pool.Unlock()
+			continue
+		} else {
+			// mark for deletion
+			// NOTE: Once a Connection can no longer take a new request
+			// it never will be able to again. Therefore a conn will always go down the delete
+			// chan. Repeats down the chan are handled in the del method
+			pool.currentConn.obj.ReleaseUsage()
+			pool.delConnCh <- pool.currentConn
+			// after marking for deletion, move on to next connection
+			pool.currentConn = pool.currentConn.next
+		}
+
+		pool.Unlock()
+	}
+}
+
+// Get returns a ConnPoolObject which has been called with AcquireUsage at least once
+// it is the callers responsibility to release it's interest after usage by calling ReleaseUsage()
+// TODO: Allow set timeout
+func (pool *connPool) Get() ConnPoolObject {
+	conn := <-pool.goodConnCh
+	for conn.obj.Expended() {
+		conn = <-pool.goodConnCh
+	}
+	return conn.obj
+}
+
+/*
+func (hc *connNode) RoundTrip(r *http.Request) (*http.Response, error) {
```

Comment:
let's clean-up this file and commented-out code before merging the PR

Reply:
02659031f77425438f77200a5ca5e699e36451d5